xwiki-platform-core/xwiki-platform-panels/xwiki-platform-panels-api/src/main/java/org/xwiki/panels/internal/LeftPanelsUIExtensionManager.java

```diff
@@ -35,7 +38,7 @@
 public class LeftPanelsUIExtensionManager extends AbstractPanelsUIExtensionManager
 {
     @Override
-    protected String getConfiguration()
+    protected List<String> getConfiguration()
```

Comment:
If you have to do it it means you introduced a breaking change on XWikiPreference side. I guess it means that any code which manipulate/request panels configuration is going to be broken.

Reply:
Could you give me an example? The class is internal and the method protected so I didn't thought there could be a breaking change outside the package.

Reply:
What I said is that you had to modify this method (which is not a breaking change) BECAUSE you introduce a breaking change: the list of plugins is not stored the same way anymore.

Reply:
I've change the type of the `XWiki.XWikiPreferences > leftPanels` field, from String to Page, but the way it is stored doesn't change. In this case the relational storage is set to false which mean a [`StringListProperty`](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/objects/classes/ListClass.java#L414) will be created instead of a [`StringProperty`](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/objects/classes/StringClass.java#L83) but AFAIK they are stored in the same way in the database (i.e. as a string where values are separated by a comma or something else). I could have avoid using a `List<String>` and just convert the list to a string [here](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-panels/xwiki-platform-panels-api/src/main/java/org/xwiki/panels/internal/RightPanelsUIExtensionManager.java#L41).
Though, I'm not very aware of what could be broken (other than internal code) after changing `String` fields to `Page` fields. Maybe you can give me some examples?

Reply:
StringListProperty yes but the separator changed. StringListProperty uses `|` as separator while leftPanels separator used to be based on `,`.

Reply:
I can probably force the separator to be `,` instead of `|` but I need to know how I can make sure no breaking changes have been made.

Reply:
Well in theory this change should break all panels in XWiki Standard since they are listed with a `,`: https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-distribution/xwiki-platform-distribution-ui/xwiki-platform-distribution-ui-base/src/main/resources/XWiki/XWikiPreferences.xml#L2739

Reply:
Anyway even if you fix the separator (which might not be as easy as it sounds) there is still the breakage illustrated by your getConfiguration change: getValue() now return a List while any code which manipulate this field expect a String.

Reply:
It seems that the separator used is a space ` `, but still my panels are not broken and they are well listed:
![image](https://user-images.githubusercontent.com/2213999/43726670-104e13e2-99a0-11e8-892c-1fa0300e2ac9.png)
I can set the `separators` of all multi-select Page fields to `|, ` if you think that can help.

Reply:
> It seems that the separator used is a space

Definitely not a space in https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-distribution/xwiki-platform-distribution-ui/xwiki-platform-distribution-ui-base/src/main/resources/XWiki/XWikiPreferences.xml#L2739 and https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-panels/xwiki-platform-panels-api/src/main/java/org/xwiki/panels/internal/AbstractPanelsUIExtensionManager.java#L99.

Reply:
> getValue() now return a List while any code which manipulate this field expect a String.

That's true, the value bound to the property will be a `List` instead of a `String`. There might be other cases like that so I can see at least 2 solutions:
1. Put back the type of the field to `StringClass`.
2. Keep the `PageClass` but override the [`newProperty`](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/objects/classes/ListClass.java#L407-L420) method to create a `StringProperty` in case we are using a multi-select field, without relational storage and when some new value is set to true (to use a String instead of a List). This solution is too complex I think.

Maybe you have other ideas?

> Definitely not a space

Sorry, that was the default separator for the page field, which isn't used by the string field.

Reply:
So I did some debugging and the reason why upgrade "works" is mostly because of  things:
* XWikiPreference is very special and is not really based on standard properties tables (it's custom mapped on a dedicated table) which means it's going trough BaseClass#fromValueMap instead of loading each object property based on the table
* the hack you added in https://github.com/xwiki/xwiki-platform/pull/803/files#diff-4bbd11fef5480d7e7db21e44ee739043R851. This change is actually wrong because you are mixing UI things (fromString which is supposed to deal with html form values, so UI stuff) and storage thing (fromValue) which don't always work the same way. For example by default in ListClass the UI syntax and storage syntax is not exactly the same (not the same separators plus escaping support on storage side).

Putting aside retro compatibility, instead of calling fromString in fromValue the right way to support StringListProperty in a custom mapping (other than rewritting all this mess) would be to overwrite setValue in StringListProperty (to not crash and burn when getting a String) IMO.

I agree that given the javadoc available on fromString and fromValue/fromValueMap it was not exactly obvious for you :)

Reply:
Ok thanks, I'll do the change.

Reply:
Actually I see that a getCustomMappingValue (in BaseProperty) exist so maybe a better fix would be to introduce setCustomMappingValue.